src/Generator.cpp

```diff
@@ -21,362 +26,655 @@
+                    calibratednoise[layer_id][i] = sigma_noise*mip/sampl;
+            }
+        }
+        else
+            calibratednoise = readCalibration(parameters_.generation().calib_file);
     }
+    cout << "Done !"<<endl;
 
-    if (debug) std::cout << " incident energy: " << parameters_.generation().energy << " simulated energy: " << energygen << std::endl;    
 
-    // if requested display a few events
-    if (iev<=parameters_.display().events) {
-      for (const auto& id_energy : event.hits()) { 
-        hCellEnergyEvtMap.at(id_energy.first).Reset();
-        hCellEnergyEvtMap.at(id_energy.first).Fill(id_energy.second);
-      }
-      canvas.emplace_back(display(hCellEnergyEvtMap,iev));    
+    // start main loop on all events
+    double hit_outside_geom = 0.;
+    double tot_hits = 0;
 
+    // Histograms for transverse profile
+    TH1F *hTransverseProfile[layer_max-layer_min];
+    char* hTransProfName = new char[30];
+
+    for (int i = layer_min; i < layer_max; i++) {
+        sprintf(hTransProfName, "hTransverseProfile_%d", i);
+        hTransverseProfile[i] = new TH1F(hTransProfName, "Generated transverse profile (cm)", 600, 0., 120);
     }
-    output_.fillTree(event, geometry_);
 
-  }
-  std::cout << std::endl;
-  std::cout << nevents << " events generated " << std::endl;  
-  std::cout << std::endl;
+    // Build the tree map (except for external geometry) ; one key corresponds to one layer
+    std::unordered_map<int, Tree*> tree_map;
+
+    if (parameters_.geometry().type!=Parameters::Geometry::Type::External) {
+
+        for (int layer_id = layer_min; layer_id < layer_max; layer_id++) {
+
+            std::vector<double > cell_x;
+            std::vector<double > cell_y;
+            for (Cell c : cell_collection[layer_id]) {
+                cell_x.push_back(c.getX());
+                cell_y.push_back(c.getY());
+            }
+
+            double x_min = *std::minmax_element(cell_x.begin(), cell_x.end()).first;
+            double x_max = *std::minmax_element(cell_x.begin(), cell_x.end()).second;
+            double y_min = *std::minmax_element(cell_y.begin(), cell_y.end()).first;
+            double y_max = *std::minmax_element(cell_y.begin(), cell_y.end()).second;
+
+            // Geometry build with the cell center position, add a margin for the
+            // square corners containing the boarder cells
+            x_min -= parameters_.geometry().small_cell_side;
+            x_max += parameters_.geometry().large_cell_side;
+            y_min -= parameters_.geometry().large_cell_side * 2;
+            y_max += parameters_.geometry().large_cell_side * 2;
+
+            Rectangle* plan = new Rectangle(
+                new Point((float)x_min, (float)y_max),
+                new Point((float)x_max, (float)y_min)
+            );
+
+            Tree* tree = new Tree(plan, 5);
+            float x_c, y_c;
+            double side;
+            for(Cell& c : cell_collection[layer_id]) {
```

Comment:
Use a const reference (the cells are not modified, right?)

Reply:
I didn't managed to use the const reference. The cells are distributed in the leaf od the Quadtree and the method push back didn't supported the const. There is a way to use pushback and const together, I didn't find it ?

Reply:
Maybe it should be a vector of const cells then. If the cells are not supposed to be modified then they should be const.